File: src/main/java/com/teamwizardry/wizardry/common/spell/component/ModuleShape.java
package com.teamwizardry.wizardry.common.spell.component;

import java.util.List;
import java.util.Map;

import com.teamwizardry.wizardry.api.spell.PatternShape;

import net.minecraft.item.Item;

public class ModuleShape extends Module
{
    protected final String form;
    
    public ModuleShape(PatternShape pattern, String name, List<Item> items, double baseManaCost, double baseBurnoutCost, String form, String element, Map<String, Double> modifierCosts, Map<String, List<Double>> attributeValues)
    {
        super(pattern, name, items, baseManaCost, baseBurnoutCost, element, modifierCosts, attributeValues);
        this.form = form;
    }
    
    @Override public PatternShape getPattern() { return (PatternShape) pattern; }
    
    public String getForm() { return form; }
}

File: src/main/kotlin/com/teamwizardry/wizardry/common/spell/component/ModuleShape.kt
package com.teamwizardry.wizardry.common.spell.component

import net.minecraft.item.Item

class ModuleShape(
    pattern: PatternShape,
    name: String,
    items: List<Item>,
    baseManaCost: Double,
    modifierCosts: Map<String, Double>,
    attributeValues: Map<String, List<Double>>
) : Module(pattern, name, items, baseManaCost, modifierCosts, attributeValues) {
    override val pattern: Pattern = pattern
        get() = field as PatternShape
}